graph/cassandra/quadstore.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2014 The Cayley Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cassandra
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/barakmich/glog"
+	"github.com/gocql/gocql"
+
+	"github.com/google/cayley/graph"
+	"github.com/google/cayley/graph/iterator"
+	"github.com/google/cayley/quad"
+)
+
+func init() {
+	graph.RegisterQuadStore(QuadStoreType, true, newQuadStore, createNewCassandraGraph, nil)
+}
+
+const (
+	// DefaultKeyspace is the name of the (preexisting) keyspace prepared in Cassandra.
+	DefaultKeyspace = "cayley"
+
+	// DefaultReadConsistency is the default consistency level for reads to Cassandra.
+	DefaultReadConsistency = "one"
+
+	// DefaultWriteConsistency is the default consistency level for writes to Cassandra.
+	DefaultWriteConsistency = "any"
+
+	// QuadStoreType is the string identifier for this quadstore.
+	QuadStoreType = "cassandra"
+)
+
+type QuadStore struct {
+	sess             *gocql.Session
+	size             int64
+	horizon          int64
+	writeConsistency gocql.Consistency
+}
+
+func getAllAddresses(addr string, options graph.Options) []string {
+	var out []string
+	out = append(out, strings.Split(addr, ",")...)
+	if val, ok := options.StringKey("address_list"); ok {
+		out = append(out, strings.Split(val, ",")...)
+	}
+	return out
+}
+
+func getConsistency(s string) gocql.Consistency {
+	for i, c := range gocql.ConsistencyNames {
+		if c == s {
+			return gocql.Consistency(i)
+		}
+	}
+	glog.Fatalf("No such Cassandra consistency %s", s)
+	return 0
+}
+
+func clusterWithOptions(addr string, options graph.Options) (*gocql.ClusterConfig, error) {
+	allAddrs := getAllAddresses(addr, options)
+	cluster := gocql.NewCluster(allAddrs...)
+	keyspace := DefaultKeyspace
+	if val, ok := options.StringKey("keyspace"); ok {
+		keyspace = val
+	}
+	consistencyStr := DefaultReadConsistency
+	if val, ok := options.StringKey("consistency"); ok {
+		consistencyStr = val
+	}
+	cluster.Consistency = getConsistency(consistencyStr)
+	cluster.Keyspace = keyspace
```

Comment:
a few more handy options for C* performance for gocql is controlling number of conns and streams and the discover hosts option 
cluster.NumConns = 2
cluster.NumStreams = 16
cluster..DiscoverHosts = true

discover hosts true should be safe, conns and streams can be optional params passed in